### PR TITLE
Add a design document for cached output files

### PR DESCRIPTION
**Pull request by xylar**
This is being implemented in #184

### CODE REVIEW

**Comment by xylar** [comment]
@mark-petersen, as I mentioned in #184, I think we need a design document for this work.  Here it is.

@matthewhoffman, this work is general enough that it could affect the `landice` core.  If you are not interested in this capability, that's fine. Please just give the design doc a quick glance to make sure it seems to "do no harm" for you.  If you do want to have the capability, please look more closely and make sure it meets the needs you foresee having.

I will make some additional comments or ask questions in specific sections.

**Comment by xylar** [comment]
To render the design doc in sphinx, set up the compass environment as usual.  Hopefully, you can set it up on your laptops because that makes it easier to open the results in a browser.  Once you have sourced the activation script, run:
```
cd docs
make html
```
Then, open `_build/html/design_docs/cached_outputs.html` in your browser.

File: docs/design_docs/cached_outputs.rst [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+Requirement: either "normal" or "cached" versions of a step
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2021/07/30
+
+Contributors: Xylar Asay-Davis
+
+We **do not** require the ability to set up a "normal" and a "cached" version
+of the same step within a ``compass`` test case or suite.  (If this is not the
+case, it would place important constraints on the design solution.)
```
Comment (xylar) [inline docs/design_docs/cached_outputs.rst:95]: This is the main requirement where I would like feedback.  Do we foresee needing both "normal" and "cached" version of steps?  If so, how do we avoid having a second copy of every test case where we want to have cached outputs (and every test case that depends on the "cached" test case)?  If we don't need that ability, the design I have here should serve us well.

File: docs/design_docs/cached_outputs.rst [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    # Options related to the current test case
+    [test_case]
+    
+    # The steps from the test case where output files should be downloaded
+    # from the server rather than computed
+    cached = mesh
+
+Both of these approaches assume that we always have either the "normal" or the
+"cached" version of a step within a test case or test suite and developers or
+users are free to choose between them, see :ref:`des_normal_or_cached`.
```
Comment (xylar) [inline docs/design_docs/cached_outputs.rst:154]: This is the main design choice where I could use some help brainstorming.  Can either of you think of better, cleaner ways of letting developers (e.g. via test suites) and users (e.g. via config files) determine which steps use cached outputs and which do not?

File: docs/design_docs/cached_outputs.rst [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+flexibility about which steps are cached and which are not.  It is not clear
+to me how we achieve this flexibility without requiring that a given step
+either be set up as "normal" or "cached", and not both in the same work
+directory.
+
+
+
+
+Implementation
+--------------
```
Comment (xylar) [inline docs/design_docs/cached_outputs.rst:237]: Waiting until we iterate on requirements and design before I rework the implementation in #184

File: docs/design_docs/cached_outputs.rst (hunk 2 above)
Comment (xylar) [reply docs/design_docs/cached_outputs.rst]: I came up with a better implementation.  It's reasonably efficient for adding cached outputs to test suites and for setting up test cases where *all* steps are cached.  it's not great for setting up individual test cases where only some of the steps should be cached, and we probably just want to provide test suites if such situations emerge very often.

**Comment by xylar** [comment]
@mark-petersen, I believe this is finished but please feel free to commend and request changes.

**Review by mark-petersen — APPROVED** [review]
@xylar thanks, this looks great. I really like the brief notation of 'c' in 
```
compass setup -n 40c
```
and then the ability for a longer format with the `--cached` flag.

I agree that we do not need to require the ability to set up a cached version of every step in a test case. For some domains, like low-resolution regular hexes, it does not make sense to cache the input file.

**Comment by xylar** [comment]
> I agree that we do not need to require the ability to set up a cached version of every step in a test case. For some domains, like low-resolution regular hexes, it does not make sense to cache the input file.

To be clear, the current implementation *does* support caching the outputs of every single step of every test case.  But I haven't actually run all tests and cached their outputs.  So in that sense, we will need to add cache files to the ``compass_cache`` database using the ``compass cache`` tool as we find we need them.

**Comment by xylar** [comment]
Thanks very much for the review!  I think I'll merge this, and then we can make minor tweaks in #184 if need be.

**Pull request merged**